src/app/games/lasca/LascaMove.ts

```diff
@@ -0,0 +1,131 @@
+        }
+        const vector: Coord = start.getVectorToward(end);
+        if (Math.abs(vector.x) !== 1 || Math.abs(vector.y) !== 1) {
+            return MGPFallible.failure(LascaMoveFailure.MOVE_STEPS_MUST_BE_SINGLE_DIAGONAL());
+        }
+        return MGPFallible.success(new LascaMove([start, end], true));
+    }
+    public static encoder: MoveEncoder<LascaMove> = new class extends MoveEncoder<LascaMove> {
+        public encodeMove(move: LascaMove): JSONValueWithoutArray {
+            return {
+                coords: move.coords.map((coord: Coord): JSONValueWithoutArray => {
+                    return Coord.encoder.encode(coord) as JSONValueWithoutArray;
+                }),
+                isStep: move.isStep,
+            };
+        }
+        public decodeMove(encoded: JSONValueWithoutArray): LascaMove {
+            const casted: JSONObject = encoded as JSONObject;
+            assert(casted.coords != null, 'Encoded LascaMove should have a coords field');
+            assert(casted.isStep != null, 'Encoded LascaMove should have a isStep field');
+            const encodedCoords: JSONValueWithoutArray[] =
+                Utils.getNonNullable(casted.coords) as JSONValueWithoutArray[];
+            const coords: Coord[] = encodedCoords.map((x: JSONValue) => Coord.encoder.decode(x));
+            return new LascaMove(coords, casted.isStep as boolean);
+        }
+    };
+    private constructor(private readonly coords: Coord[], public readonly isStep: boolean) {
+        super();
+    }
+    public toString(): string {
+        const coords: Coord[] = this.coords;
+        const coordStrings: string[] = coords.map((coord: Coord) => coord.toString());
+        const coordString: string = coordStrings.join(', ');
+        return 'LascaMove(' + coordString + ')';
+    }
+    public equals(other: LascaMove): boolean {
+        return this.getRelation(other) === 'EQUALITY';
+    }
+    public getRelation(other: LascaMove): 'EQUALITY' | 'PREFIX' | 'INEQUALITY' {
+        const thisLength: number = this.coords.length;
+        const otherLength: number = other.coords.length;
+        const minimalLength: number = Math.min(thisLength, otherLength);
+        for (let i: number = 0; i < minimalLength; i++) {
+            if (this.coords[i].equals(other.coords[i]) === false) return 'INEQUALITY';
+        }
+        if (thisLength === otherLength) return 'EQUALITY';
+        else return 'PREFIX';
+    }
+    public getCoord(index: number): MGPFallible<Coord> {
+        if (index < 0 || index >= this.coords.length) {
+            return MGPFallible.failure('invalid index');
+        }
+        return MGPFallible.success(this.coords[index]);
+    }
+    public getCoordsCopy(): Coord[] {
+        return ArrayUtils.copyImmutableArray(this.coords);
+    }
+    public getStartingCoord(): Coord {
+        return this.getCoord(0).get();
+    }
+    public getEndingCoord(): Coord {
+        return this.coords[this.coords.length - 1];
+    }
```

Comment:
getCoord(this.coords.length-1) ? (Imagine t'as 0 coords, getEndingCoords throw alors que t'as écrit une getCoord safe :D)

Reply:
t'as jamais zéro coords :D

Reply:
jusqu'au jour où y'a un bug et on se retrouve avec this.coords = [], et ça propage des undefined partout et on doit supprimer MGP et recommencer !

Reply:
Maintenant this.coords est un MGPSet, ça ne change rien, right :D ?

Reply:
On regroupe ça dans le ticket "créer UniqueList et l'utiliser dans Lasca" ?

Reply:
non on fixe ça avant de pourriter MGPSet dans develop ! (tpr une des plusieurs conversations à ce sujet)

Reply:
Bah, j'ai fait un Set.toList() vu que de toute façon là, l'ordre est respecté dans les Set (c'était donc déjà "pourri")
Du coup:
1. je tickette et on fait plus tard ?
2. je tickette pas car c'est impératif hygiéniquement selon toi que je le fasse de suite ?
3. je tickette pas car meh, Set.toList() est ok car les set vont pas se mettre à s'autoshuffler anyway ?

Reply:
Comment ça l'ordre était déjà respecté ? L'ordre est *pas* respecté : `MGPSet([1]).add(2)` et `MGPSet([2]).add(1)` sont les même set et donnent un .toList différent

Reply:
MGPSet([1]).add(2) et MGPSet([2]).add(1) sont les même set et donnent un .toList différent.
Si le toList est différent, l'ordre est respecté en interne, on est pas obligé mais pour l'instant, de facto, c'est respecté !
Mais les LascaStack sont des readonly array maintenant, on est bon ! TPR